docs/ref/models/fields.txt

```diff
@@ -51,6 +51,12 @@ for "no data;" the Django convention is to use the empty string, not
 and ``blank=True`` set. In this situation, ``null=True`` is required to avoid
 unique constraint violations when saving multiple objects with blank values.
 
+Additionally, be cautious of depending on :attr:`~Field.null` to enforce a
+"not null" constraint at the database level for string-based fields, as
+databases may accept the empty string as a value despite :attr:`~Field.null`
```

Comment:
I think "may accept" needs more precision. I guess it relates to the note below about Oracle storing null if you pass it an empty string? It would be interesting to know if any other databases (outside of Django's built-in backends) do that.

Reply:
Hi @timgraham,

Thank you for the review! I understand your concern about the precision of the "may accept".

Should I clarify that among Django's built-in database backends PostgreSQL, MySQL, and SQLite, consider empty strings as valid string values? (The case for Oracle is already documented in the note on ``Field.null``.)

Reply:
There are more hints about the empty string behavior than I first thought

We could change this paragraph to be:
```
That ``null=False`` has no impact on whether empty strings are allowed for
string-based fields. If you wish to prevent empty strings being saved to the
database, you need to call ``full_clean()`` before saving a model object when
:attr:`blank=False <Field.blank>`. You may also want to set :attr:`default=None
<Field.default>`, as the Django convention is to use empty strings by default.
```
The "as the Django convention is to use empty strings by default" is also said above but I think it's useful to reiterate it

Reply:
Hello @sarahboyce! Sorry, I saw this after I had made and pushed my most recent commits. I am considering replacing the admonition I just included in favour of this. What do you think? Should I replace the advice with CheckConstraint in favour of this? Myself I prefer the ``CheckConstraint`` over calling full-clean() on every save.

Reply:
I read the admonition. I'm not sure we can say what "the preferred method" is as there are different ways to approach this
Maybe just adding ```Empty strings can still be saved to the database when ``null=False``.``` at the end of the paragraph about string based fields is enough?

Reply:
Okay I'll do that. Thanks.

Reply:
On second thought, I think we still have to add some pointers for how one might ensure that empty strings are not saved in database so readers know where to look for solutions. Here is what I propose:
After the second paragraph:
```reST
Empty strings can still be saved to the database when ``null=False``. If you
wish to disallow them, one method is to use a :class:`CheckConstraint`.
Alternatively, you could add extra validation logic to the model's
:meth:`~Model.clean()` method.
```
Let me know what you think.

Reply:
Been thinking about this some more and I think maybe it's enough to move the comment of `the Django convention is to use the empty string, not NULL` further up so

```
Avoid using :attr:`~Field.null` on string-based fields such as
:class:`CharField` and :class:`TextField`. The Django convention is to use an
empty string, not ``NULL`` as the "no data" state for string-based fields. If a
string-based field has ``null=True``, that means it has two possible values for
"no data": ``NULL``, and the empty string. In most cases, it's redundant to have
two possible values for "no data". One exception is when a :class:`CharField`
has both ``unique=True`` and ``blank=True`` set. In this situation,
``null=True`` is required to avoid unique constraint violations when saving
multiple objects with blank values.
```

And update
```diff
 .. attribute:: Field.null
 
-If ``True``, Django will store empty values as ``NULL`` in the database. Default
-is ``False``.
+If ``True``, Django can store ``NULL`` values in the database. Default is
+``False``.
```
as that might be slightly more acurate

Reply:
I think the paragraph with "the Django convention" you propose is indeed more effective than the previous one. But I'm not so sure about
```diff
+If ``True``, Django can store ``NULL`` values in the database. Default is
+``False``.
```
Even for string-based fields (as far as I have been able to check) if ``null=True`` Django will in fact save ``NULL`` in the database for "no data". So maybe we ought to let that part alone?
The inconsistencies only happens when ``null=False``, which is where our problem lies

Reply:
Yes sorry :+1: makes sense

Reply:
How about modifying the passage on string fields to:

```diff
Avoid using :attr:`~Field.null` on string-based fields such as
:class:`CharField` and :class:`TextField`. The Django convention is to use an
empty string, not ``NULL`` as the "no data" state for string-based fields.
+ If a string based-field has ``null=False`` empty strings can still be saved.
If a string-based field has ``null=True``, that means it has two possible values
for "no data": ``NULL``, and the empty string. In most cases, it's redundant to
have two possible values for "no data". One exception is when a
:class:`CharField` has both ``unique=True`` and ``blank=True`` set. In this
situation, ``null=True`` is required to avoid unique constraint violations when
saving multiple objects with blank values.
```